item.lua

```diff
@@ -0,0 +1,294 @@
+-----------------------------------------------------------------------
+item = {}
+
+--- @param tab table<any,fun(item:item):boolean>
+function item.condition_burrow(tab, burrow, outside)
+    if outside then
+        addCondition(
+            tab,
+            function (item) return not containsItem(burrow,item) end
+        )
+    else
+        addCondition(
+            tab,
+            function (item) return containsItem(burrow,item) end
+        )
+    end
+end
+
+function item.condition_type(tab, type)
+    addCondition(
+        tab,
+        function(item)
+            return df.item_type[item:getType()] == string.upper(type)
+        end
+    )
+end
+
+function item.condition_reachable(tab)
+    local cgroups = citizenWalkabilityGroups()
+    addCondition(tab, function (item) return fastReachable(item,cgroups) end)
+end
+
+function item.condition_unreachable(tab)
+    local cgroups = citizenWalkabilityGroups()
+    addCondition(tab, function (item) return not fastReachable(item,cgroups) end)
+end
+
+-- uses the singular form without stack size (i.e., prickle berry)
+-- does not work for corpse pieces like "wool"
+function item.condition_description(tab, desc)
+    addCondition(
+        tab,
+        function(item) return dfhack.items.getDescription(item, 1) == desc end
+    )
+end
+
+function item.condition_material(tab, material)
+    addCondition(
+        tab,
+        function(item) return dfhack.matinfo.decode(item):toString() == material end
+    )
+end
+
+function item.condition_wear(tab, level, upper)
+    if upper then
+        addCondition(tab, function(item) return item.wear <= level end)
+    else
+        addCondition(tab, function(item) return item.wear >= level end)
+    end
+end
+
+function item.condition_quality(tab, levelst, upper)
+    local level = argparse.nonnegativeInt(levelst, 'wear')
+    if upper then
+        addCondition(tab, function(item) return item.quality <= level end)
+    else
+        addCondition(tab, function(item) return item.quality >= level end)
+    end
+end
+
+function item.act(action,conditons,options)
+    local count = 0
+    local changed = 0
+    local invisible = 0
+    for _, item in pairs(df.global.world.items.other.IN_PLAY) do
+        -- never act on items used for constructions/building materials and carried by hostiles
+        -- also skip artifacts, unless explicitly told to include them
+        if  item.flags.construction or
+            item.flags.in_building or
+            item.flags.hostile or
+            (item.flags.artifact and not options.artifact)
+        then
+            goto skipitem
+        end
+
+        -- check conditions provided via options
+        for _, condition in pairs(conditons) do
+            if not condition(item) then goto skipitem end
+        end
+
+        -- only try to melt things that are actually meltable
+        -- TOTHINK: treated as implicit filter, should we count the item?
```

Comment:
I'm thinking that maybe we should add more action-implied filters (e.g. don't forbid already forbidden items) and perform this check before the potentially more expensive user-provided filters. This would make those actions provide lower counts than running `item count` with the same filters. Maybe a `--dry-run` option is in order.

Reply:
I implemented `--dry-run.`